example/jupyterlab/README.md

```diff
@@ -0,0 +1,251 @@
+- Small fixes to allow configuration via the JupyterLab advanced settings
+- Nested columns names using multi-part dot seperated paths
+- Support for spaces in nested columns names (backtiks each segment)
+- Visually distinct icons for `table`, `column`, `alias`, `functions` suggestions
+- Added support for code-completion of
+	- Functions including descriptions
+	- Table alias
+	- Table alias followed by dot
+	- Partially typed columns including multi-part fields (needed insertText)
+    - Map and Array subscripts `map_col['key'].<tab>`, `array_col[0].<tab>`
+
+
+- no support for fully qualified table names `databaseName.schemaName.tableName` (might be something we want to support)
+
+
+
+
+# JupyterLab Integration
+
+## Install
+
+JupyterLab supports `sql-language-server`. More details here
+https://jupyterlab-lsp.readthedocs.io/en/latest/Language%20Servers.html
+
+All that is required is for sql-language-server to be installed.
+
+```bash
+$ npm install --save-dev sql-language-server
+```
+
+When you create a `%%sql` cell it will trigger the initialization of the `sql-language-server`. You can verify that it starts correctly by looking at the status bar.
+
+![initialized](https://github.com/cccs-jc/sql-language-server/blob/develop/example/jupyterlab/images/initialized-screenshot.png)
+
+
+
+# sqlite3
+
+In order to execute `%%sql` cells you need to install the IPython sql magic.
+
+```bash
+$ pip3 install ipython-sql
+```
+
+In order to code complete your SQL statement you need to configure `sql-language-server` using JupyterLab's Advanced Settings.
+
+> Advanced Settings Editor -> Language Server
+
+Assuming you have an sqlite3 database file at `/tmp/demodb.sqlite` enter the following configuration:
+```json
+{
+    "language_servers": {
+        "sql-language-server": {
+            "serverSettings": {
+                "sqlLanguageServer": {
+                    "connections": [
+                        {
+                            "name": "sqlite-conf",
+                            "adapter": "sqlite3",   
+                            "filename": "/tmp/demodb.sqlite"
+                        }
+                    ]
+                }
+            }
+        }
+    }
+}
+```
+
+The code cell now supports syntax highlighting, code completion and execution of SQL statements.
+
+![sqlite](https://github.com/cccs-jc/sql-language-server/blob/develop/example/jupyterlab/images/jupyterlab-sqlite-demo.gif)
+
+
+
+# Spark
+
+Sql-language-server now supports a forth `adapter` named `json` which lets you provide your own schema for code completion to the sql-language-server. This is handy for scenarios like Spark where drivers are not easily integrated into a JavaScript process.
+
+The idea is, a user triggers the genration of the schema by calling a function (work in progress). This might be a IPython magic for example
+
+```
+%sql-language-server --generateSchema
```

Comment:
@cccs-jc Do you have a plan to work to implement this `generateSchema` command?

Reply:
@cccs-jc I understand this generating is gonna be an important thing that this integration goes well.
Also this is gonna be core feature of sql-language-server itself.  If you're busy, I'm also happy to have a plan to implement this. 😄

Reply:
I have a proof of concept that does the generation of the schema https://github.com/cccs-jc/sparksql_editor

This project is derived from the sparksql_magic project which enables execution of spark SQL statements in a JupyterLab notebook. I have extended it with many more features. The sparksql_editor generates the schema for sql-language-server.

I'm also adding support for TrinoDB and using the same approach where the python magic produces a schema file that the sql-language-server uses for code completion.

sparksql_editor is not finished and I might actually rename it to reflect that it support more than just spark. Other big data SQL engine support should be relatively easy to add to the project.

Reply:
@cccs-jc wow, cool. That's brilliant and I'm so excited!